server/tools.go

```diff
@@ -0,0 +1,414 @@
+package server
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+	gotmpl "text/template"
+
+	jsonv2 "github.com/go-json-experiment/json"
+	jsontext "github.com/go-json-experiment/json/jsontext"
+
+	"github.com/ollama/ollama/api"
+)
+
+type State int
+
+// TODO: potentially coalesce states
+const (
+	SendTokens State = iota
+	GreedyToolWithPrefix
+	GreedyToolNoPrefix
+	ForceTools
+	ToolSuffix
+	ContainsPrefix
+	PartialPrefix
+	NotPartialPrefix
+	Done
+)
+
+type ExternalState int
+
+const (
+	ToolCallFound ExternalState = iota
+	ToolCallSendPartial
+	ToolCallAccumulate
+	ToolCallSendTokens
+)
+
+func (s ExternalState) String() string {
+	switch s {
+	case ToolCallFound:
+		return "ToolCallFound"
+	case ToolCallSendPartial:
+		return "ToolCallSendPartial"
+	case ToolCallAccumulate:
+		return "ToolCallAccumulate"
+	case ToolCallSendTokens:
+		return "ToolCallSendTokens"
+	default:
+		return fmt.Sprintf("Unknown ExternalState (%d)", s)
+	}
+}
+
+func (s State) String() string {
+	switch s {
+	case SendTokens:
+		return "SendTokens"
+	case GreedyToolWithPrefix:
+		return "GreedyToolWithPrefix"
+	case GreedyToolNoPrefix:
+		return "GreedyToolNoPrefix"
+	case ForceTools:
+		return "ForceTools"
+	case ToolSuffix:
+		return "ToolSuffix"
+	case PartialPrefix:
+		return "PossiblePrefix"
+	case Done:
+		return "Done"
+	case ContainsPrefix:
+		return "PartialPrefix"
+	default:
+		return fmt.Sprintf("Unknown State (%d)", s)
+	}
+}
+
+// TODO: simplify if possible
+type ToolParser struct {
+	tmpl        *gotmpl.Template
+	state       State
+	sb          *strings.Builder
+	toolPrefix  string
+	toolIndex   int
+	ParserState ExternalState
+	Done        bool
+}
+
+// ? move to a separate file
+// parseJSONToolCalls attempts to parse a JSON string into a slice of ToolCalls.
+// Returns parsed tool calls, a boolean indicating if the JSON is incomplete, and a boolean indicating if the tool calls were found
+func (p *ToolParser) parseJSONToolCalls(s string) ([]api.ToolCall, bool, bool) {
```

Comment:
might break into smaller funcs and have general helpers for use in diff places. not sure if that is helpful

Reply:
I think that make sense too. It's close but it should be easier to follow at a glance